components/codec/buffer.rs

```diff
@@ -0,0 +1,435 @@
+            for _ in 0..100 {
+                base_write.push(rand::random());
+            }
+
+            let mut buffer = base.clone();
+            let mut buffer = buffer.as_mut_slice();
+
+            buffer.bytes_mut(13)[..13].clone_from_slice(&base_write[0..13]);
+            assert_eq!(&buffer[0..13], &base_write[0..13]);
+            assert_eq!(&buffer[13..], &base[13..]);
+            buffer.advance_mut(13);
+
+            // Acquire 10, only write 5.
+            buffer.bytes_mut(10)[..5].clone_from_slice(&base_write[13..18]);
+            assert_eq!(&buffer[0..5], &base_write[13..18]);
+            assert_eq!(&buffer[5..], &base[18..]);
+            buffer.advance_mut(5);
+
+            buffer.bytes_mut(22)[..22].clone_from_slice(&base_write[18..40]);
+            assert_eq!(&buffer[0..22], &base_write[18..40]);
+            assert_eq!(&buffer[22..], &base[40..]);
+            buffer.advance_mut(22);
+
+            assert_eq!(buffer, &base[40..]);
+        }
+    }
+
+    #[test]
+    fn test_buffer_writer_vec() {
+        unsafe {
+            // A series of bytes to write
+            let mut base_write: Vec<u8> = Vec::with_capacity(100);
+            for _ in 0..100 {
+                base_write.push(rand::random());
+            }
+
+            let mut buffer: Vec<u8> = Vec::with_capacity(20);
+            buffer.bytes_mut(13)[..13].clone_from_slice(&base_write[0..13]);
+            buffer.advance_mut(13);
+            assert_eq!(&buffer[0..13], &base_write[0..13]);
+            assert_eq!(buffer.len(), 13);
+
+            // Vec remaining 7, acquire 10, only write 5
+            assert!(buffer.bytes_mut(10).len() >= 10);
+            buffer.bytes_mut(10)[..5].clone_from_slice(&base_write[13..18]);
+            buffer.advance_mut(5);
+            assert_eq!(&buffer[0..18], &base_write[0..18]);
+            assert_eq!(buffer.len(), 18);
+
+            // Reset len
+            buffer.set_len(7);
+            buffer.bytes_mut(16)[..16].clone_from_slice(&base_write[18..34]);
+            buffer.advance_mut(16);
+            assert_eq!(&buffer[0..7], &base_write[0..7]);
+            assert_eq!(&buffer[7..23], &base_write[18..34]);
+            assert_eq!(buffer.len(), 23);
+
+            buffer.bytes_mut(15)[..15].clone_from_slice(&base_write[34..49]);
+            buffer.advance_mut(15);
+            assert_eq!(&buffer[0..7], &base_write[0..7]);
+            assert_eq!(&buffer[7..38], &base_write[18..49]);
+            assert_eq!(buffer.len(), 38);
+
+            buffer.bytes_mut(2)[..2].clone_from_slice(&base_write[49..51]);
+            buffer.advance_mut(2);
+            assert_eq!(&buffer[0..7], &base_write[0..7]);
+            assert_eq!(&buffer[7..40], &base_write[18..51]);
+            assert_eq!(buffer.len(), 40);
+        }
+    }
+
+    /// Test whether it is safe to store values in `Vec` after `len()`, i.e. during
+    /// reallocation these values are copied.
+    #[test]
+    fn test_vec_reallocate() {
```

Comment:
Seems like an implementation-dependent behavior, the caller should advance after write if it wants to make sure that the written data will be copied on reallocation?

Reply:
Yes, I understand the purpose of this test. What I concern is that maybe we should require the caller to advance instead of relying on the assumption.

Reply:
Hmmm. If the caller does not do advance, then what it wrote is just not visible and.. equally it does nothing.